### PR TITLE
Add RPC marshaling support for custom interfaces

### PR DESCRIPTION
**Pull request by matteo-prosperi**
This PR addresses https://github.com/microsoft/vs-streamjsonrpc/issues/774.

I added the `RpcMarshalable` attribute that can be applied to an interface that
- extends `IDisposable`
- doesn't have properties
- doesn't have events

I added unit tests and docs.

### CODE REVIEW

File: src/StreamJsonRpc/JsonMessageFormatter.cs [hunk 1]
```diff
@@ -1479,34 +1457,29 @@ public override void WriteJson(JsonWriter writer, Stream? value, JsonSerializer
...
+            private string DebuggerDisplay => $"Converter for marhalable interfaces";
 
             public override bool CanConvert(Type objectType) =>
-                objectType == this.implicitlyConvertedType ||
-                (this.implicitlyConvertedType.IsGenericTypeDefinition && objectType.IsGenericType && objectType.GetGenericTypeDefinition() == this.implicitlyConvertedType);
+                MessageFormatterRpcMarshaledContextTracker.GetMarshalOptionsForType(objectType) is not null;
 
             public override object? ReadJson(JsonReader reader, Type objectType, object? existingValue, JsonSerializer serializer)
             {
+                (JsonRpcProxyOptions ProxyOptions, JsonRpcTargetOptions TargetOptions)? marshalOptions
```
Comment (matteo-prosperi) [inline src/StreamJsonRpc/JsonMessageFormatter.cs:1474]: The current implementation is simpler than the original one which had multiple instances of `RpcMarshalableImplicitConverter`, one for each marshalable type. On the other hand, we are not caching the options anymore and we have to evaluate the `objectType` twice (one during `CanConvert` and then again during `ReadJson`). Should I go back to caching the options in different instances of `RpcMarshalableImplicitConverter` (one per type)?

File: src/StreamJsonRpc/RpcMarshalableAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace StreamJsonRpc
+{
+    using System;
+
+    /// <summary>
+    /// Marks an interface as marshalable.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Interface, AllowMultiple = false, Inherited = false)]
```
Comment (matteo-prosperi) [inline src/StreamJsonRpc/RpcMarshalableAttribute.cs:11]: Note: this can only be applied to interfaces. And is not automatically inherited..

File: src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs [hunk 3]
```diff
@@ -87,6 +87,44 @@ private enum MarshalMode
...
+                {
+                    throw new NotSupportedException(Resources.MarshalableInterfaceHasEvents);
+                }
+
+                if (type.GetProperties().Any())
+                {
+                    throw new NotSupportedException(Resources.MarshalableInterfaceHasProperties);
+                }
+
+                return (new JsonRpcProxyOptions(), new JsonRpcTargetOptions { NotifyClientOfEvents = false, DisposeOnDisconnect = true });
```
Comment (matteo-prosperi) [inline src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs:122]: @AArnott, what should we do for `DisposeOnDisconnect` for custom interfaces? I opted for `true` considering that we are forcing the interface to extend `IDisposable`, but I don't have a strong opinion on this.

File: test/StreamJsonRpc.Tests/MarshalableProxyTests.cs [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+using Nerdbank.Streams;
+using Newtonsoft.Json;
+using StreamJsonRpc;
+using Xunit;
+using Xunit.Abstractions;
+
+/// <summary>
+/// Tests the proxying of interfaces marked with <see cref="RpcMarshalableAttribute"/>.
+/// </summary>
+public abstract class MarshalableProxyTests : TestBase
```
Comment (matteo-prosperi) [inline test/StreamJsonRpc.Tests/MarshalableProxyTests.cs:21]: I replicated all the tests that are available for `IDisposable` and added some making sure that
- `DisposeOnDisconnect` is honored
- the proxies can be used
- exceptions are thrown if the interfaces are invalid for marshaling (non disposable, have events, or have properties)
- generic interfaces work

File: src/StreamJsonRpc/RpcMarshalableAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (AArnott) [inline src/StreamJsonRpc/RpcMarshalableAttribute.cs:1]: nit: all the other attribute classes are declared in files under the `Reflection` folder. But the namespace you have is correct.

File: src/StreamJsonRpc/RpcMarshalableAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace StreamJsonRpc
```
Comment (AArnott) [inline src/StreamJsonRpc/RpcMarshalableAttribute.cs:4]: nit: consider using a namespace statement instead of a block:

```cs
namespace StreamJsonRpc;
```

File: src/StreamJsonRpc/Resources.resx [hunk 7]
```diff
@@ -208,6 +208,15 @@
   <data name="JsonRpcMethodAndIgnoreAttributesFound" xml:space="preserve">
     <value>Conflicting JsonRpcMethodAttribute and JsonRpcIgnoreAttribute found on the same method: {0}.</value>
   </data>
+  <data name="MarshalableInterfaceHasEvents" xml:space="preserve">
+    <value>Interfaces with events cannot be marshaled</value>
```
Comment (AArnott) [inline src/StreamJsonRpc/Resources.resx:212]: Please end each of your error messages with periods.

File: src/StreamJsonRpc/JsonMessageFormatter.cs [hunk 8]
```diff
@@ -1479,34 +1457,29 @@ public override void WriteJson(JsonWriter writer, Stream? value, JsonSerializer
...
-                this.targetOptions = targetOptions;
-            }
-
-            private RpcMarshalableImplicitConverter(RpcMarshalableImplicitConverter copyFrom, Type implicitlyConvertedType)
-                : this(implicitlyConvertedType, copyFrom.jsonMessageFormatter, copyFrom.proxyOptions, copyFrom.targetOptions)
-            {
             }
 
-            private string DebuggerDisplay => $"Implicit converter for: {this.implicitlyConvertedType.Name}";
+            private string DebuggerDisplay => $"Converter for marhalable interfaces";
```
Comment (AArnott) [inline src/StreamJsonRpc/JsonMessageFormatter.cs:1467]: typo: marhalable

File: src/StreamJsonRpc/JsonMessageFormatter.cs (hunk 1 above)
Comment (AArnott) [reply src/StreamJsonRpc/JsonMessageFormatter.cs:1474]: I think what you have here is fine, so long as if it's too slow to look up the options for a given type, we can optimize the `GetMarshalOptionsForType` method itself to cache the result.

File: src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs [hunk 9]
```diff
@@ -87,6 +87,44 @@ private enum MarshalMode
             MarshallingRealObject = 1,
         }
 
+        internal static (JsonRpcProxyOptions ProxyOptions, JsonRpcTargetOptions TargetOptions)? GetMarshalOptionsForType(Type type)
+        {
+            Requires.NotNull(type, nameof(type));
```
Comment (AArnott) [inline src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs:92]: No need to test for null inputs for internal methods since we're using nullable-type reference annotations which would produce a build warning (which would break PR/CI builds).

File: src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs [hunk 10]
```diff
@@ -87,6 +87,44 @@ private enum MarshalMode
...
+            }
+
+            if (type.IsInterface && type.GetCustomAttribute<RpcMarshalableAttribute>() is not null)
+            {
+                if (typeof(IDisposable).IsAssignableFrom(type) is false)
+                {
+                    throw new NotSupportedException(Resources.MarshalableInterfaceNotDisposable);
+                }
+
+                if (type.GetEvents().Any())
```
Comment (AArnott) [inline src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs:112]: using the `.Any()` extension method is slower than using the `Length` or `Count` properties, and sometimes allocates memory when the property alternatives would not. For each of these, as they return arrays can you use the `Length` property instead?

File: src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs [hunk 11]
```diff
@@ -87,6 +87,44 @@ private enum MarshalMode
...
+                if (implicitlyMarshaledType == type ||
+                    (implicitlyMarshaledType.IsGenericTypeDefinition &&
+                     type.IsConstructedGenericType &&
+                     implicitlyMarshaledType == type.GetGenericTypeDefinition()))
+                {
+                    return (proxyOptions, targetOptions);
+                }
+            }
+
+            if (type.IsInterface && type.GetCustomAttribute<RpcMarshalableAttribute>() is not null)
```
Comment (AArnott) [inline src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs:105]: The happy path through this block allocates 2 arrays and potentially many objects. It seems worth caching any Type that passes through the happy path so that each compliant one only pays the penalty once. 
We could also reuse the resulting tuple to avoid a couple more allocations (effectively making this method alloc-free after the first call for a given Type) since although they are mutable, this is an internal method and its callers are not expected to mutate the returned objects.

File: test/StreamJsonRpc.Tests/MarshalableProxyTests.cs [hunk 12]
```diff
@@ -0,0 +1,522 @@
...
+    {
+    }
+
+    [RpcMarshalable]
+    public interface IMarshalableWithProperties : IDisposable
+    {
+        int Foo { get; }
+    }
+
+    [RpcMarshalable]
```
Comment (AArnott) [inline test/StreamJsonRpc.Tests/MarshalableProxyTests.cs:71]: What about a negative test for an interface that *isn't* attributed?

File: test/StreamJsonRpc.Tests/MarshalableProxyTests.cs [hunk 13]
```diff
@@ -0,0 +1,522 @@
...
+        await disposed.WaitAsync(this.TimeoutToken);
+        Assumes.NotNull(this.server.ReceivedProxy);
+
+        WeakReference weakProxy = new WeakReference(this.server.ReceivedProxy);
+        this.server.ReceivedProxy = null;
+        return (weakProxy, weakTarget);
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private async Task<WeakReference> NoLeakWhenTransmissionFailsAfterTokenGenerated_Helper()
```
Comment (AArnott) [inline test/StreamJsonRpc.Tests/MarshalableProxyTests.cs:349]: Such helper methods *do* need to be separate methods, but now that C# allows us to attribute local methods, the most maintainable and reader-friendly way to lay this out is to move all the helper methods into their test methods as local methods (typically named something shorter as well, like `Helper`).

File: test/StreamJsonRpc.Tests/MarshalableProxyTests.cs [hunk 14]
```diff
@@ -0,0 +1,522 @@
...
+    }
+
+    [RpcMarshalable]
+    public interface IMarshalable : IDisposable
+    {
+        Task DoSomethingAsync();
+    }
+
+    [RpcMarshalable]
+    public interface IGenericMarshalable<T> : IMarshalable
```
Comment (AArnott) [inline test/StreamJsonRpc.Tests/MarshalableProxyTests.cs:55]: How about a negative test for a non-attributed interface that derives from an attributed interface?

File: test/StreamJsonRpc.Tests/MarshalableProxyTests.cs [hunk 15]
```diff
@@ -0,0 +1,522 @@
...
+
+        public Task AcceptNonDisposableMarshalableAsync(INonDisposableMarshalable nonDisposable) => Task.CompletedTask;
+
+        public Task AcceptMarshalableWithPropertiesAsync(IMarshalableWithProperties marshalableWithProperties) => Task.CompletedTask;
+
+        public Task AcceptMarshalableWithEventsAsync(IMarshalableWithEvents marshalableWithEvents) => Task.CompletedTask;
+    }
+
+    [DataContract]
+    public class ProxyContainer<T>
```
Comment (AArnott) [inline test/StreamJsonRpc.Tests/MarshalableProxyTests.cs:445]: How about a class which implements a marshalable interface and includes `[DataContract]` but is passed by its concrete type rather than as the interface? In such a case, it should be passed by-value instead of by-reference.

File: test/StreamJsonRpc.Tests/MarshalableProxyTests.cs [hunk 16]
```diff
@@ -0,0 +1,522 @@
...
+        public Task<int> AcceptDataContainerAsync(DataContainer dataContainer) => Task.FromResult(dataContainer.Data?.Value ?? 0);
+
+        public Task AcceptNonDisposableMarshalableAsync(INonDisposableMarshalable nonDisposable) => Task.CompletedTask;
+
+        public Task AcceptMarshalableWithPropertiesAsync(IMarshalableWithProperties marshalableWithProperties) => Task.CompletedTask;
+
+        public Task AcceptMarshalableWithEventsAsync(IMarshalableWithEvents marshalableWithEvents) => Task.CompletedTask;
+    }
+
+    [DataContract]
```
Comment (AArnott) [inline test/StreamJsonRpc.Tests/MarshalableProxyTests.cs:444]: How about a class which implements a marshalable interface that *isn't* attributed with `[DataContract]` (which I expect to work)?

**Review by AArnott — CHANGES_REQUESTED** [review]
Looks good overall. It's exciting to see this come together so quickly.

File: test/StreamJsonRpc.Tests/MarshalableProxyTests.cs (hunk 15 above)
Comment (matteo-prosperi) [reply test/StreamJsonRpc.Tests/MarshalableProxyTests.cs:445]: `IMarshalableDataAsArg_ShouldSerialize` and `IMarshalableDataAsObjectWithinArg_ShouldSerialize` should already test that. We were already testing this case for `IDisposable` and I duplicated all the `IDisposable` tests here.

File: src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs (hunk 3 above)
Comment (AArnott) [reply src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs:122]: I think that's a good default.

File: test/StreamJsonRpc.Tests/MarshalableProxyTests.cs (hunk 15 above)
Comment (AArnott) [reply test/StreamJsonRpc.Tests/MarshalableProxyTests.cs:445]: Ah yes, I see. Thanks.

File: src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs [hunk 17]
```diff
@@ -87,6 +90,50 @@ private enum MarshalMode
...
+
+                    if (type.GetProperties().Length > 0)
+                    {
+                        throw new NotSupportedException(Resources.MarshalableInterfaceHasProperties);
+                    }
+
+                    return RpcMarshalableInterfaceDefaultOptions;
+                }
+
+                return null;
```
Comment (matteo-prosperi) [inline src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs:133]: @AArnott I added caching but should I cache `null`?
The only case in which we could have `null` here if I get an interface here but it is not marked as marshalable and it's not in the implicit list. Should I instead throw an exception if I ever encounter such interface? Or is there a valid use case (or a future use case) for an interface to be passed here that is not marshalable? (the only case I can think of is if there is a lower priority formatter that will also accept interfaces)

Comment (matteo-prosperi) [reply src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs]: Not caching `null` anymore as discussed in person.

File: src/StreamJsonRpc/JsonMessageFormatter.cs [hunk 18]
```diff
@@ -1477,31 +1461,58 @@ public override void WriteJson(JsonWriter writer, Stream? value, JsonSerializer
         }
 
         [DebuggerDisplay("{" + nameof(DebuggerDisplay) + "}")]
-        private class RpcMarshalableImplicitConverter : JsonConverter
+        private class RpcMarshalableReadConverter : JsonConverter
```
Comment (matteo-prosperi) [inline src/StreamJsonRpc/JsonMessageFormatter.cs:1464]: @AArnott, I split `RpcMarshalableImplicitConverter` into two.
`RpcMarshalableReadConverter` resolves the type options during deserialization, but cannot write. This is added to the formatter's converter list.
`RpcMarshalableConverter` has properties for the interface type and the corresponding options (like the old `RpcMarshalableImplicitConverter` did). This is necessary because `WriteJson` doesn't take the definition type and trying to find the type at runtime would misbehave in case an object implements multiple marshalable interfaces (this is what was breaking the tests on Mono).

File: src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs [hunk 19]
```diff
@@ -87,6 +90,62 @@ private enum MarshalMode
...
+                    proxyOptions = typeProxyOptions;
+                    targetOptions = typeTargetOptions;
+                    MarshaledTypes.TryAdd(type, (proxyOptions, targetOptions));
+                    return true;
+                }
+            }
+
+            if (type.GetCustomAttribute<RpcMarshalableAttribute>() is not null)
+            {
+                if (typeof(IDisposable).IsAssignableFrom(type) is false)
```
Comment (BertanAygun) [inline src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs:125]: Do we need to check for default interface methods, can we even check for it? I was thinking it could be trouble if client/server has slightly different versions where the default implementation is different. Although it wouldn't be a problem specific to marshaled interface case.

Comment (matteo-prosperi) [inline src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs:125]: @AArnott, should I allow `IAsyncDisposable` or `Microsoft.VisualStudio.Threading.IAsyncDisposable` as an alternative to `IDisposable`?

File: src/StreamJsonRpc/Reflection/RpcMarshalableAttribute.cs [hunk 20]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace StreamJsonRpc;
+
+using System;
+
+/// <summary>
+/// Marks an interface as marshalable.
+/// </summary>
+[AttributeUsage(AttributeTargets.Interface, AllowMultiple = false, Inherited = false)]
+public class RpcMarshalableAttribute : Attribute
```
Comment (matteo-prosperi) [inline src/StreamJsonRpc/Reflection/RpcMarshalableAttribute.cs:12]: @AArnott, should we have `DisposeTargetOnDisconnect` as a property of this attribute? If we do, should the default be `true` or `false`?

File: src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs (hunk 19 above)
Comment (matteo-prosperi) [reply src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs:125]: Discussed in person. Any consideration regarding default methods in interfaces should be addressed across both marshaling interfaces and "normal" server/client proxies as they should behave in the same way.

File: src/StreamJsonRpc/Reflection/RpcMarshalableAttribute.cs (hunk 20 above)
Comment (AArnott) [reply src/StreamJsonRpc/Reflection/RpcMarshalableAttribute.cs:12]: I don't think we should. The only behavior should be `true`. My thinking is that the underlying protocol already describes that disposing the proxy _will_ dispose the marshaled object. This means that passing a marshalable object over RPC signifies an ownership transfer. A dropped connection should not leave any resources unfreed, so we should guarantee that Dispose is called on behalf of the proxy if the proxy cannot do it do to a dropped connection.

File: src/StreamJsonRpc/Reflection/RpcMarshalableAttribute.cs [hunk 21]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace StreamJsonRpc;
+
+using System;
+
+/// <summary>
+/// Marks an interface as marshalable.
```
Comment (AArnott) [inline src/StreamJsonRpc/Reflection/RpcMarshalableAttribute.cs:9]: I feel like this could be beefed up a bit. How about:

> Designates an interface that is used in an RPC contract to marshal the object so the receiver can invoke remote methods on it instead of serializing the object to send its data to the remote end.

Then a `<remarks>` section that has something like this would probably be helpful:

```cs
/// <remarks>
/// <see href="https://github.com/microsoft/vs-streamjsonrpc/blob/main/doc/rpc_marshalable_objects.md">Learn more about marshable interfaces</see>.
/// </remarks>
```

File: test/StreamJsonRpc.Tests/MarshalableProxyTests.cs [hunk 22]
```diff
@@ -0,0 +1,621 @@
...
+    [Fact]
+    public async Task InterfacesMustBeMarkedAsRpcMarshalable()
+    {
+        await Assert.ThrowsAnyAsync<Exception>(() => this.client.AcceptNonMarshalableAsync(new NonDataContractMarshalable()));
+    }
+
+    [Fact]
+    public async Task RpcMarshalableAttributeDoesntAffectDerivedInterfaces()
+    {
+        await Assert.ThrowsAnyAsync<Exception>(() => this.client.AcceptNonMarshalableDerivedFromMarshalablesAsync(new NonDataContractMarshalable()));
```
Comment (AArnott) [inline test/StreamJsonRpc.Tests/MarshalableProxyTests.cs:195]: Why does this throw? 
What if the interface included attributes to instruct the formatter as to how to format it as data, the way interfaces can typically do? I would expect that to work, so it would be interesting to see a test to show it.

File: src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs (hunk 19 above)
Comment (AArnott) [reply src/StreamJsonRpc/Reflection/MessageFormatterRpcMarshaledContextTracker.cs:125]: I don't think so. Proxies never implement these async dispose interfaces because Dispose is always transmitted to the server as a notification, so there is nothing for the proxy to wait for. I think the marshaling behavior is that the server object can implement the async disposable interfaces and if it does, StreamJsonRpc will invoke it, but this doesn't manifest on the proxy. As this interface becomes what the proxy implements, it shouldn't derive from the async interface.

File: test/StreamJsonRpc.Tests/MarshalableProxyTests.cs (hunk 22 above)
Comment (matteo-prosperi) [reply test/StreamJsonRpc.Tests/MarshalableProxyTests.cs:195]: You are saying that an interface could be marked as `[DataContract]`? If that is supported, I will add tests for that.

But what do we do in case an interface is both `[DataContract]` and `RpcMarshalable`? Which one wins?

About why this throws: `AcceptNonMarshalableDerivedFromMarshalablesAsync` takes a parameter of type `INonMarshalableDerivedFromMarshalable` where
```cs
    public interface INonMarshalableDerivedFromMarshalable : IMarshalable
    {
    }

    [RpcMarshalable]
    public interface IGenericMarshalable<T> : IMarshalable
    {
        Task<T> DoSomethingWithParameterAsync(T paremeter);
    }
```

We don't allow to use a derived interface type as if it were also `RpcMarshalable`.

This is reinforced by `RpcMarshalableAttribute` being marked as `Inherited = false`.

Comment (matteo-prosperi) [reply test/StreamJsonRpc.Tests/MarshalableProxyTests.cs:238]: We clarified in person that an interface could be both `[RpcMarshalable]` and otherwise serializable (e.g., through `[MessagePackFormatter]` or `[JsonConverter]`) and that the desired behavior would be either throwing an exception or marshaling.

I changed the implementation to marshal these interfaces (I fear that the check necessary to throw an exception would lead to negative performance impact).

For MessagePack, the solution is simply to move `RpcMarshalableImplicitResolver` from the bottom to the top of the resolver list.

For Newtonsoft.Json, the solution is to leverage the contract resolver.

**Review by AArnott — APPROVED** [review]
Looks really good. Just a few touch-up suggestions.

File: src/StreamJsonRpc/JsonMessageFormatter.cs [hunk 23]
```diff
@@ -1515,26 +1491,17 @@ public override void WriteJson(JsonWriter writer, object? value, JsonSerializer
                 {
                     writer.WriteNull();
                 }
+                else if (this.interfaceType.IsAssignableFrom(value.GetType()) is false)
```
[the remark this answers is not in the document]
Comment (matteo-prosperi) [reply src/StreamJsonRpc/JsonMessageFormatter.cs:1494]: Will change to `!`.
Myself, I never liked the `!` syntax and even preferred `== false` to it because the `!` at the beginning of the statement is sometimes hard to see. Also `is false` resembles natural language more.
But will adapt to the local codestyle and use `!` here :-)

File: src/StreamJsonRpc/MessagePackFormatter.cs [hunk 24]
```diff
@@ -587,6 +587,9 @@ private MessagePackSerializerOptions MassageUserDataOptions(MessagePackSerialize
             // Add our own resolvers to fill in specialized behavior if the user doesn't provide/override it by their own resolver.
             var resolvers = new IFormatterResolver[]
             {
+                // Support for marshalled objects.
+                new RpcMarshalableImplicitResolver(this),
```
[the remark this answers is not in the document]
Comment (matteo-prosperi) [reply src/StreamJsonRpc/MessagePackFormatter.cs:591]: If I switch the order of the resolvers, an interface that is both marshalable and serializable (e.g. through `MessagePackFormatterAttribute`) would be serialized:
1. it would be a different behavior than when using Newtonsoft.Json
2. we discussed in [this comment](https://github.com/microsoft/vs-streamjsonrpc/pull/777#discussion_r825001683) that marshalability should have precedence. For reference:
> `RpcMarshable` should win because it's universal (same attribute regardless of formatter), and switching formatters shouldn't change a type from serializing to marshalling

File: doc/rpc_marshalable_objects.md [hunk 25]
```diff
@@ -0,0 +1,244 @@
...
+
+1. The receiver calls `IDisposable.Dispose()` on the proxy.
+1. The JSON-RPC connection is closed.
+
+## Protocol
+
+The protocol for proxying a disposable object is based on [general marshaled objects](general_marshaled_objects.md).
+
+The responsibility to release resources is on the receiver of the proxy.
+That is, when the proxy holder calls either of these methods, it should also send the `$/releaseMarshaledObject` message back to the target object owner. When the proxy holder is using StreamJsonRpc, this is handled automatically when `IDisposable.Dispose()` is invoked on the proxy.
```
[the remark this answers is not in the document]
Comment (matteo-prosperi) [reply doc/rpc_marshalable_objects.md:245]: It was copied from [observer.md](https://github.com/microsoft/vs-streamjsonrpc/blob/main/doc/observer.md#protocol), but I don't think we need this statement,
```
The responsibility to release resources is on the receiver of the proxy.
```
should be sufficient.

File: src/StreamJsonRpc/MessagePackFormatter.cs (hunk 24 above)
[the remark this answers is not in the document]
Comment (AArnott) [reply src/StreamJsonRpc/MessagePackFormatter.cs]: I get that if you move it too far down it would be serialized, but the user-types resolver is typically empty until they populate it IIRC. So I thought one position down wouldn't have broken this.

[the remark this answers is not in the document]
Comment (matteo-prosperi) [reply src/StreamJsonRpc/MessagePackFormatter.cs]: Moving it one position down breaks the case I described above because `userSuppliedOptions.Resolver` by default includes `AttributeFormatterResolver`.

[the remark this answers is not in the document]
Comment (AArnott) [reply src/StreamJsonRpc/MessagePackFormatter.cs]: Ok. We'll leave it then.
